x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/ElementType.java

```diff
@@ -8,37 +8,46 @@
 package org.elasticsearch.compute.data;
 
 import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.TransportVersions;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.util.Maps;
+
+import java.io.IOException;
+import java.util.Map;
 
 /**
  * The type of elements in {@link Block} and {@link Vector}
  */
 public enum ElementType {
-    BOOLEAN("Boolean", BlockFactory::newBooleanBlockBuilder),
-    INT("Int", BlockFactory::newIntBlockBuilder),
-    LONG("Long", BlockFactory::newLongBlockBuilder),
-    FLOAT("Float", BlockFactory::newFloatBlockBuilder),
-    DOUBLE("Double", BlockFactory::newDoubleBlockBuilder),
+    BOOLEAN(0, "Boolean", BlockFactory::newBooleanBlockBuilder, BooleanBlock::readFrom),
+    INT(1, "Int", BlockFactory::newIntBlockBuilder, IntBlock::readFrom),
+    LONG(2, "Long", BlockFactory::newLongBlockBuilder, LongBlock::readFrom),
+    FLOAT(3, "Float", BlockFactory::newFloatBlockBuilder, FloatBlock::readFrom),
+    DOUBLE(4, "Double", BlockFactory::newDoubleBlockBuilder, DoubleBlock::readFrom),
     /**
      * Blocks containing only null values.
      */
-    NULL("Null", (blockFactory, estimatedSize) -> new ConstantNullBlock.Builder(blockFactory)),
+    NULL(5, "Null", (blockFactory, estimatedSize) -> new ConstantNullBlock.Builder(blockFactory), BlockStreamInput::readConstantNullBlock),
 
-    BYTES_REF("BytesRef", BlockFactory::newBytesRefBlockBuilder),
+    BYTES_REF(6, "BytesRef", BlockFactory::newBytesRefBlockBuilder, BytesRefBlock::readFrom),
 
     /**
      * Blocks that reference individual lucene documents.
      */
-    DOC("Doc", DocBlock::newBlockBuilder),
+    DOC(7, "Doc", DocBlock::newBlockBuilder, in -> { throw new UnsupportedOperationException("can't read doc blocks"); }),
 
     /**
      * Composite blocks which contain array of sub-blocks.
      */
-    COMPOSITE("Composite", BlockFactory::newAggregateMetricDoubleBlockBuilder),
+    COMPOSITE(8, "Composite", BlockFactory::newAggregateMetricDoubleBlockBuilder, CompositeBlock::readFrom),
 
     /**
      * Intermediate blocks which don't support retrieving elements.
      */
-    UNKNOWN("Unknown", (blockFactory, estimatedSize) -> { throw new UnsupportedOperationException("can't build null blocks"); });
+    UNKNOWN(9, "Unknown", (blockFactory, estimatedSize) -> { throw new UnsupportedOperationException("can't build null blocks"); }, in -> {
```

Comment:
Minute point: for future extensibility, maybe space out (multiple of two) the writeable code instead of using consecutive numbers: e.g.:
0 - null
1 - unknown
2-3 - unused
4-15: java primitives (including those not supported yet such as byte)
16-32: rest of the objects (doc, composite, etc..)